### PR TITLE
Simplify `Inject`/`Interpret` for 1-field records

### PR DESCRIPTION
**Pull request by Gabriel439**
Fixes https://github.com/dhall-lang/dhall-haskell/issues/346

Now a Haskell type like:

```haskell
data Example = Foo Bool | Bar Text
```

... corresponds to this Dhall type:

```dhall
{ Foo : Bool | Bar : Text }
```

### CODE REVIEW

File: dhall/tests/Dhall/Test/Dhall.hs [hunk 1]
```diff
@@ -169,48 +169,48 @@ deriving instance Interpret ()
...
     , E0 `shouldInjectInto` "< E0 | E1 | E2 >.E0"
     , E1 `shouldInjectInto` "< E0 | E1 | E2 >.E1"
     , E2 `shouldInjectInto` "< E0 | E1 | E2 >.E2"
 
-    , M0 True `shouldInjectInto` "(< M0 : { _1 : Bool } | M1 | M2 : { _1 : {} } >).M0 { _1 = True }"
-    , M1 `shouldInjectInto` "(< M0 : { _1 : Bool } | M1 | M2 : { _1 : {} } >).M1"
-    , M2 () `shouldInjectInto` "(< M0 : { _1 : Bool } | M1 | M2 : { _1 : {} } >).M2 { _1 = {=} }"
+    , M0 True `shouldInjectInto` "(< M0 : Bool | M1 | M2 >).M0 True"
+    , M1 `shouldInjectInto` "(< M0 : Bool | M1 | M2 >).M1"
+    , M2 () `shouldInjectInto` "(< M0 : Bool | M1 | M2 >).M2"
```
Comment (sjakobi) [inline dhall/tests/Dhall/Test/Dhall.hs]: Could you extend `Mixed` to have a constructor with multiple unnamed fields, like `M3 Text Natural` (and `M4 Bool ()`)?

Also a union with named fields would be interesting, e.g. `data UnionWithSelectors = A | B { b1 :: Text, b2 :: Bool }`.

**Comment by Profpatsch** [comment]
Ah great, I was very surprised by the implicit `_1`, `_2`, … fields.

**Comment by patrickmn** [comment]
I think this broke a place where I'm deserializing a one-field record:

```haskell
data MyRecord = MyRecord { foo :: Text }
```

```dhall
myRecord = { foo : Text }
```

-->
```
Error: Expression doesn't match annotation

{ myRecord : ?
              - Text
              + { ? : ? }
...
```

Just making sure that's intentional.

**Comment by Gabriel439** [comment]
@patrickmn: It was intentional, but we can make this behavior configurable via `InterpretOptions`

**Comment by patrickmn** [comment]
I understand it now, the error was just a bit puzzling until I looked at the recent merges.

My data type was just a placeholder, so it's no bother :)

**Comment by Gabriel439** [comment]
@patrickmn: I decided to disable the breaking change by default, at least for the upcoming release, by guarding it behind an option.  See: https://github.com/dhall-lang/dhall-haskell/pull/1321

**Pull request merged**